src/Simplex/Messaging/Server/NtfStore.hs

```diff
@@ -1,19 +1,76 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
 {-# LANGUAGE OverloadedStrings #-}
 
 module Simplex.Messaging.Server.NtfStore where
 
+import Control.Concurrent.STM
+import Control.Monad (foldM)
+import Data.Int (Int64)
+import qualified Data.Map.Strict as M
 import Data.Time.Clock.System (SystemTime (..))
 import qualified Simplex.Messaging.Crypto as C
 import Simplex.Messaging.Encoding.String
 import Simplex.Messaging.Protocol (EncNMsgMeta, MsgId, NotifierId)
+import Simplex.Messaging.TMap (TMap)
+import qualified Simplex.Messaging.TMap as TM
 
-data MsgNtf = MsgNtf MsgId SystemTime C.CbNonce EncNMsgMeta
+newtype NtfStore = NtfStore (TMap NotifierId (TVar [MsgNtf]))
+
+data MsgNtf = MsgNtf
+  { ntfMsgId :: MsgId,
+    ntfTs :: SystemTime,
+    ntfNonce :: C.CbNonce,
+    ntfEncMeta :: EncNMsgMeta
+  }
+
+storeNtf :: NtfStore -> NotifierId -> MsgNtf -> IO ()
+storeNtf (NtfStore ns) nId ntf = do
+  TM.lookupIO nId ns >>= atomically . maybe newNtfs (`modifyTVar'` (ntf :))
+  -- TODO coalesce messages here once the client is updated to process multiple messages
+  -- for single notification.
+  -- when (isJust prevNtf) $ incStat $ msgNtfReplaced stats
+  where
+    newNtfs = TM.lookup nId ns >>= maybe (TM.insertM nId (newTVar [ntf]) ns) (`modifyTVar'` (ntf :))
+
+deleteNtfs :: NtfStore -> NotifierId -> IO ()
+deleteNtfs (NtfStore ns) nId = atomically $ TM.delete nId ns
+
+flushNtfs :: NtfStore -> NotifierId -> IO [MsgNtf]
+flushNtfs (NtfStore ns) nId = do
+  TM.lookupIO nId ns >>= maybe (pure []) swapNtfs
+  where
+    swapNtfs v =
+      readTVarIO v >>= \case
+        [] -> pure []
+        -- if notifications available, atomically swap with empty array
+        _ -> atomically (swapTVar v [])
+
+deleteExpiredNtfs :: NtfStore -> Int64 -> IO Int
+deleteExpiredNtfs (NtfStore ns) old =
+  foldM (\expired -> fmap (expired +) . expireQueue) 0 . M.keys =<< readTVarIO ns
+  where
+    expireQueue nId = TM.lookupIO nId ns >>= maybe (pure 0) expire
+    expire v = readTVarIO v >>= \case
+      [] -> pure 0
+      _ -> atomically $ do
+        ntfs <- readTVar v
+        case reverse ntfs of
+          -- check the last message first, it is the earliest
+          MsgNtf {ntfTs} : _ | systemSeconds ntfTs < old -> do
+            let !ntfs' = filter (\MsgNtf {ntfTs = ts} -> systemSeconds ts >= old) ntfs
+            writeTVar v ntfs'
+            pure $! length ntfs - length ntfs'
+          _ -> pure 0
```

Comment:
why not just filter first, then reverse only those written to TVar

smth like

```
let !ntfs' = filter (\MsgNtf {ntfTs = ts} -> systemSeconds ts >= old) ntfs
if null ntfs'
  then
    pure 0
  else do
    writeTVar v (reverse ntfs')
    pure $! length ntfs - length ntfs'
```

that first element is being specially checked here doesn't seem to optimize anything as it requires reversing first anyway

Reply:
reverse is much faster operation than filter, and in this case it is executed lazily - that is, it only scans to the last element and takes it - it could just use `last`, as it's known the list is not empty.